File: Money.java
import java.io.IOException;
import java.nio.file.Files;
import java.nio.file.Paths;
import java.util.*;
import java.util.concurrent.*;

public class Money extends Thread {

	private static ArrayList<HashMap<String, Object>> customers = new ArrayList<HashMap<String, Object>>();
	private static ArrayList<HashMap<String, String>> banks = new ArrayList<HashMap<String, String>>();
	private static ArrayList<String> threadList = new ArrayList<String>();
	private static BlockingQueue<HashMap<String, String>> bqMsg;
	private static int bankCount = 0;
	private String name;
	
	public Money(String name) {
		this.name = name;
	}

	public static void main(String[] args) {
		try {
			createBanks();
			createCustomers();
			bqMsg = new ArrayBlockingQueue<HashMap<String, String>>(20);
			createThreads();
		} catch (Exception ex) {
			System.out.println(ex.toString());
		}
	}
	
	public static HashMap<String, Object> getCustomer(String customerName) {
		HashMap<String, Object> customer = new HashMap<String, Object>();
		for (HashMap<String, Object> cust : customers) {
			if(cust.get("name").equals(customerName)) {
				return cust;
			}
		}
		return customer;
	}
	
	public static HashMap<String, String> getBank(String bankName) {
		HashMap<String, String> bank = new HashMap<String, String>();
		for (HashMap<String, String> b : banks) {
			if(b.get("name").equals(bankName)) {
				return b;
			}
		}
		return bank;
	}
	
	public static int getRandomAmount(String customerName) {
		HashMap<String, Object> customer = getCustomer(customerName);
		int max;
		if (Integer.parseInt(((String) customer.get("amount")).trim()) >= 50)
			max = 50;
		else
			max = Integer.parseInt((String) customer.get("amount"));
		
		Random ran = new Random();
		int amount = ran.nextInt(max) + 1;
		return amount;
	}
	
	public static String getRandomBank(String customerName) {
		HashMap<String, Object> customer = getCustomer(customerName);
		ArrayList<String> bankList =  (ArrayList<String>) customer.get("banks");
		return bankList.get(new Random().nextInt(bankList.size()));
	}
	
	public static boolean checkIfCustomerSendRequest(String customerName) {
		for(HashMap<String, String> request : bqMsg) {
			if(request.containsKey(customerName)) {
				return true;
			}
		}
		return false;
	}
	
	public static boolean checkIfCustomerSendRequestGranted(String customerName) {
		for(HashMap<String, String> request : bqMsg) {
			for(String value : request.values()) {
				if(value.contains("grant to " + customerName)) {
					return true;
				}
			}
		}
		return false;
	}
	
	public static boolean alreadyRequestedMoney(String customerName) {
		for(HashMap<String, String> request : bqMsg) {
			for(String key : request.keySet()) {
				if(key.equals(customerName)) {
					return true;
				}
			}
		}
		return false;
	}
	
	public static boolean checkIfCustomerSendRequestRejected(String customerName) {
		for(HashMap<String, String> request : bqMsg) {
			for(String value : request.values()) {
				if(value.contains("reject to " + customerName)) {
					return true;
				}
			}
		}
		return false;
	}
	
	public static boolean checkIfBankMoneyRequest(String bankName) {
		for(HashMap<String, String> request : bqMsg) {
			for(String value : request.values()) {
				if(value.contains("to " + bankName)) {
					return true;
				}
			}
		}
		return false;
	}
	
	public static boolean canGrantMoney(String bankName, int amount) {
		HashMap<String, String> bank = getBank(bankName);
		if(Integer.parseInt(bank.get("amount")) >= amount) {
			return true;
		}
		return false;
	}
	
	public static void grantMoney(String bankName, int amount, HashMap<String, String> request) {
		HashMap<String, String> bank = getBank(bankName);
		HashMap<String, String> grantRequest = new HashMap<String, String>();
		
		banks.remove(bank);
		Integer oldMoney = Integer.parseInt(bank.get("amount"));
		Integer moneyLeft = oldMoney - amount;
		bank.put("amount", moneyLeft.toString());
		banks.add(bank);
		
		grantRequest.put(bankName, "Request " + amount + " grant to " +  request.keySet().toArray()[0]);
		System.out.println(bankName + ": " + amount + " grant to " +  request.keySet().toArray()[0]);
		bqMsg.remove(request);
		bqMsg.add(grantRequest);
	}
	
	public static void rejectMoney(String bankName, int amount, HashMap<String, String> request) {
		HashMap<String, String> bank = getBank(bankName);
		HashMap<String, String> rejectRequest = new HashMap<String, String>();
		
		rejectRequest.put(bankName, "Request " + amount + " reject to " +  request.keySet().toArray()[0]);
		System.out.println(bankName + ": " + amount + " reject to " +  request.keySet().toArray()[0]);
		bqMsg.remove(request);
		bqMsg.add(rejectRequest);
	}
	
	public static boolean sendRequestForMoney(String customerName) {
		HashMap<String, String> request = new HashMap<String, String>();
		HashMap<String, Object> customer = getCustomer(customerName);
		
		if(Integer.parseInt(((String) customer.get("amount")).trim()) > 0 && !((ArrayList<String>)customer.get("banks")).isEmpty()) {
			int amount = Money.getRandomAmount(customerName);
			String bankName = Money.getRandomBank(customerName);
			request.put(customerName, " request " + amount + " to " + bankName);
			System.out.println(customerName + " request " + amount + " to " + bankName);
			bqMsg.add(request);
			return true;
		}
		else if(((ArrayList<String>)customer.get("banks")).isEmpty()) {
			System.out.println(customerName + ": no banks remaining, amount left : " + customer.get("amount"));
			return false;
		}
		else {
			System.out.println(customerName + " objective completed");
			return false;
		}
	}
	
	public static int getAmountRequested(String bankName) {
		for(HashMap<String, String> request : bqMsg) {
			for(String value : request.values()) {
				if(value.contains("to " + bankName)) {
					return Integer.parseInt(value.replaceAll("[\\D]", ""));
				}
			}
		}
		return 0;
	}
	
	public static void customerMoneyGranted(String customerName) {
		HashMap<String, Object> customer = getCustomer(customerName);
		HashMap<String, String> customerRequest = new HashMap<String, String>();
		
		for(HashMap<String, String> request : bqMsg) {
			for(String value : request.values()) {
				if(value.contains("grant to " + customerName)) {
					customerRequest = request;
					break;
				}
			}
		}
		
		Integer oldAmount = Integer.parseInt(((String) customer.get("amount")).trim());
		Integer amountGranted = Integer.parseInt(customerRequest.values().toString().replaceAll("[\\D]", "").trim());
		
		customers.remove(customer);
		customer.put("amount", Integer.toString(oldAmount - amountGranted));
		customers.add(customer);
		
		bqMsg.remove(customerRequest);
	}
	
	public static void customerMoneyRejected(String customerName) {
		HashMap<String, Object> customer = getCustomer(customerName);
		HashMap<String, String> customerRequest = new HashMap<String, String>();
		
		for(HashMap<String, String> request : bqMsg) {
			for(String value : request.values()) {
				if(value.contains("reject to " + customerName)) {
					customerRequest = request;
					break;
				}
			}
		}
		
		ArrayList<String> newBankList = new ArrayList<String>();
		
		for(String bank : (ArrayList<String>)customer.get("banks")) {
			if(!bank.equals(customerRequest.keySet().toArray()[0])) {
				newBankList.add(bank);
			}
		}
		
		customers.remove(customer);
		customer.put("banks", newBankList);
		customers.add(customer);
		
		bqMsg.remove(customerRequest);
	}
	
	public static void processAllBankRequests(String bankName) {
		for(HashMap<String, String> request : bqMsg) {
			if(request.values().toString().contains("to " + bankName)) {
				int amount = Integer.parseInt(request.values().toString().replaceAll("[\\D]", ""));
				
				if(Money.canGrantMoney(bankName, amount)){
					Money.grantMoney(bankName, amount, request);
				}
				else {
					Money.rejectMoney(bankName, amount, request);
				}
			}
		}
	}

	public void run() {
		
		ArrayList<String> customerList = new ArrayList<String>();
		
		for (HashMap<String, Object> cust : customers) {
			customerList.add((String) cust.get("name"));
		}
		
		ArrayList<String> bankList = new ArrayList<String>();
		for (HashMap<String, String> bank : banks) {
			bankList.add((String) bank.get("name"));
		}
		
		String currentThreadName = Thread.currentThread().getName();
		boolean isCustomerThread = customerList.contains(this.name);
		boolean restartThread = true;
		
		synchronized (bqMsg) {
			try {
				Thread.sleep(100);
			} catch (InterruptedException e) {
				e.printStackTrace();
			}
				// Customer Side
			if(isCustomerThread) {
				for (int j = 0; j < threadList.size(); j++) {
					if (this.name.equals(threadList.get(j))) {
						
						// Money granted by Bank
						if(Money.checkIfCustomerSendRequestGranted(this.name)) {
							Money.customerMoneyGranted(this.name);
						}
						
						// Money rejected by Bank
						if(Money.checkIfCustomerSendRequestRejected(this.name)) {
							Money.customerMoneyRejected(this.name);
						}
						
						if(!Money.checkIfCustomerSendRequestGranted(this.name) 
								&& !Money.checkIfCustomerSendRequestRejected(this.name)
								&& !Money.alreadyRequestedMoney(this.name)) {
							restartThread = Money.sendRequestForMoney(this.name);
						}
						
					}
				}
			}
				
			// Bank Side
			if(!isCustomerThread) {
				for (int k = 0; k < threadList.size(); k++) {
					if (this.name.equals(threadList.get(k))) {
						// check if any request for current bank
						if(Money.checkIfBankMoneyRequest(this.name)) {
							Money.processAllBankRequests(this.name);
						}
					}
				}
			}
		}
	
		if(Thread.activeCount() <= Money.bankCount + 1) {
			restartThread = false;
		}
		
		if(Thread.activeCount() <= 2) {
			System.out.println("\n\n\n** Finishing **");
			for (HashMap<String, String> bank : banks) {
				System.out.println("Bank: " + bank.get("name") + " amount left: " + bank.get("amount"));
			}
		}
		
		if(restartThread) {
			Thread threadStart = new Money(this.name);
			threadStart.setName(currentThreadName);
			threadStart.start();
		}
	}

	private static void createThreads() {
		for (String str : threadList) {
			Thread threadStart = new Money(str);
			threadStart.start();
			threadStart.setName(str);
		}
	}
	
	public static void createBanks() throws IOException {
		List<String> allLines = Files
				.readAllLines(Paths.get("banks.txt"));
		System.out.println("** Banks and financial resources **\n");

		for (String line : allLines) {
			line = line.replace("{", "").replace("}", "").replace(".", "");
			Map<String, String> bankMap = new HashMap<String, String>();
			bankMap.put("name", line.split(",")[0]);
			bankMap.put("amount", line.split(",")[1]);
			banks.add((HashMap<String, String>) bankMap);
			threadList.add(line.split(",")[0]);
			System.out.println(line.split(",")[0] + " : " + line.split(",")[1]);
			bankCount++;
		}

	}
	
	public static void createCustomers() throws IOException {
		List<String> allLines = Files
				.readAllLines(Paths.get("customers.txt"));
		System.out.println("\n\n\n** Customers and loan objectives **\n");

		ArrayList<String> bankList = new ArrayList<String>();

		for (HashMap<String, String> bank : banks) {
			bankList.add((String) bank.get("name"));
		}

		for (String line : allLines) {
			line = line.replace("{", "").replace("}", "").replace(".", "");
			Map<String, Object> customerMap = new HashMap<String, Object>();
			customerMap.put("name", line.split(",")[0]);
			customerMap.put("amount", line.split(",")[1]);
			customerMap.put("banks", bankList);
			customers.add((HashMap<String, Object>) customerMap);
			threadList.add(line.split(",")[0]);
			System.out.println(line.split(",")[0] + " : " + line.split(",")[1]);
		}
		System.out.println("\n\n\n");

	}
}

File: money.erl
-module(money).
-export([start/0, create_customers/3, create_banks/2]).

create_customers([H|Tail], CustomerEts, BankList) ->
	{Name, Amount} = H,
	io:fwrite("~s: ~w~n", [Name, Amount]),
	PID = spawn(customer, listen, [CustomerEts]),
	ets:insert(CustomerEts, {Name, Amount, PID, BankList}),
	PID ! {request, self(), Name},
	create_customers(Tail, CustomerEts, BankList);
create_customers([], CustomerEts, BankList) -> 
	io:fwrite("~n~n~n"),
	ok.

create_banks([H|Tail], BankEts) ->
	{Name, Amount} = H,
	io:fwrite("~s: ~w~n", [Name, Amount]),
	PID = spawn(bank, listen, [BankEts]),
	ets:insert(BankEts, {Name, Amount, PID}),
	PID ! {startBank},
	create_banks(Tail, BankEts);
create_banks([], BankEts) ->
	ok.

listen(BankEts, CustomerEts) ->
	receive
		{customerRequest, Name} ->
			request_amount(Name, BankEts, CustomerEts),
			listen(BankEts, CustomerEts);
		{amountGranted, Name, Amount, RequestedBank} ->
			send_amount_granted(CustomerEts, Name, Amount, RequestedBank),
			listen(BankEts, CustomerEts);
		{amountRejected, Name, Amount, RequestedBank} ->
			send_amount_rejected(CustomerEts, Name, Amount, RequestedBank),
			listen(BankEts, CustomerEts);
		{customerComplete, CustomerPID, Name} ->
			kill_customer(CustomerPID, Name),
			listen(BankEts, CustomerEts);
		{customerNoBank, CustomerPID, Name} ->
			kill_customer_no_bank(CustomerPID, Name, CustomerEts),
			listen(BankEts, CustomerEts)
	after 2500 -> 
		kill_main_process(BankEts)
	end.

get_bank_list([H|Tail], Result) ->
	{Name, Amount} = H,
	get_bank_list(Tail, lists:append(Result, [Name]));
get_bank_list([], Result) ->
	Result.
	
start() ->
	% MasterPID = spawn(money, start_banking, []),
	% keep_main_runnining(MasterPID).
	{_, Customers} = file:consult("./customers.txt"),
	{_, Banks} = file:consult("./banks.txt"),
	io:fwrite("~n~n** Banks and financial resources **~n"),
	BankEts = ets:new(bankTable,[ordered_set, public]),
	create_banks(Banks, BankEts),
	CustomerEts = ets:new(customerTable,[ordered_set, public]),
	io:fwrite("~n~n~n** Customers and loan objectives **~n"),
	create_customers(Customers, CustomerEts, get_bank_list(Banks, [])),
	listen(BankEts, CustomerEts).

% keep_main_runnining(MasterPID) ->
% 	MasterAlive = is_process_alive(MasterPID),
% 	if
% 		MasterAlive ->
% 			keep_main_runnining(MasterPID);
% 		true ->
% 			ok
% 	end.

get_random_bank(BankEts, BankList) ->
	Bank = lists:nth(rand:uniform(length(BankList)), BankList),
	[[_, PID]] = ets:match(BankEts, {Bank,'$1', '$2'}),
	{Bank, PID}.

request_amount(Name, BankEts, CustomerEts) ->
	[[AmountRequired, _, BankList]] = ets:match(CustomerEts, {Name,'$1', '$2', '$3'}),
	timer:sleep(10 * rand:uniform(10)),
	if
		AmountRequired =< 50 ->
			Amount = rand:uniform(AmountRequired);
		true ->
			Amount = rand:uniform(50)
	end,
	{RequestedBank, PID} = get_random_bank(BankEts, BankList),
	io:fwrite("~s requests a loan of ~w from ~p ~n", [Name, Amount, RequestedBank]),
	PID ! {requestAmount, self(), Name, Amount, RequestedBank}.

send_amount_granted(CustomerEts, Name, Amount, RequestedBank) ->
	io:fwrite("~s granted loan of ~p to ~p ~n", [RequestedBank, Amount, Name]),
	[[_, CustomerPID, _]] = ets:match(CustomerEts, {Name,'$1', '$2', '$3'}),
	CustomerPID ! {moneyGranted, self(), Amount}.

send_amount_rejected(CustomerEts, Name, Amount, RequestedBank) ->
	io:fwrite("~s rejected loan of ~p to ~p ~n", [RequestedBank, Amount, Name]),
	[[_, CustomerPID, _]] = ets:match(CustomerEts, {Name,'$1', '$2', '$3'}),
	CustomerPID ! {moneyRejected, self(), RequestedBank}.

kill_customer(CustomerPID, Name) ->
	io:fwrite("~s Objective Completed ~n", [Name]),
	exit(CustomerPID, "Objective Completed").

kill_customer_no_bank(CustomerPID, Name, CustomerEts) ->
	[[AmountLeft, _, _]] = ets:match(CustomerEts, {Name,'$1', '$2', '$3'}),
	io:fwrite("~s has no bank left to request money. Money left: ~w~n", [Name, AmountLeft]),
	exit(CustomerPID, "No bank left").

show_bank_info([H|T]) ->
	{Bank, AmountLeft, _} = H,
	io:fwrite("Bank: ~s has money left: ~w~n", [Bank, AmountLeft]),
	show_bank_info(T);
show_bank_info([]) ->
	ok.

kill_main_process(BankEts) ->
	io:format("~n~n~nMaster received no replies for 2.5 seconds, ending...~n~n~n"),
	BankList = ets:tab2list(BankEts),
	show_bank_info(BankList).


% start_banking() ->
% 	{_, Customers} = file:consult("./customers.txt"),
% 	{_, Banks} = file:consult("./banks.txt"),
% 	io:fwrite("~n~n** Banks and financial resources **~n"),
% 	BankEts = ets:new(bankTable,[ordered_set, public]),
% 	create_banks(Banks, BankEts),
% 	CustomerEts = ets:new(customerTable,[ordered_set, public]),
% 	io:fwrite("~n~n~n** Customers and loan objectives **~n"),
% 	create_customers(Customers, CustomerEts, get_bank_list(Banks, [])),
% 	listen(BankEts, CustomerEts).